### PR TITLE
Sane Tensorboarrd model structure visualization via name scopes

### PR DESCRIPTION
**Pull request by jakebian**
Use [tensorflow name scopes](https://www.tensorflow.org/versions/r0.11/how_tos/graph_viz/index.html#name-scoping-and-nodes) to produce drastically better model architecture visualization in tensorboard (when using the tensorboard callback).

Here's the result of this PR on the [mnist-cnn](https://github.com/fchollet/keras/blob/master/examples/mnist_cnn.py) example

**Before**
before

*Zoom in a little*
screenshot from 2016-10-31 01-51-23

**After**

screenshot from 2016-10-31 01-49-44

@fchollet let me know if you can think of a cleaner approach to this!

### CODE REVIEW

**Comment by kilotaras** [comment]
That looks awesome! Maybe replace `with K.name_scope('optimizer'):` with actual optimizer name.

File: keras/backend/theano_backend.py [hunk 1]
```diff
@@ -23,6 +23,20 @@
...
+
+    def __enter__(self):
+        return None
+
+    def __exit__(self, a, b, c):
+        return None
+
+def name_scope(name_scope):
+    # method does not apply to theano, just a placeholder
+    return EmptyContext()
```
Comment (fchollet) [inline keras/backend/theano_backend.py]: Why not make it work for Theano instead?

**Comment by fchollet** [comment]
I think we should support this for sure. But please make sure that names are consistent in both backends first.

**Comment by jakebian** [comment]
Thanks @fchollet I'll take a look at how to cleanly prepend namespaces to variable names in Theano when I get a chance.

**Comment by tboquet** [comment]
@jakebian cool work! For Theano it could be interesting to verify that the names are working with [d3viz](http://deeplearning.net/software/theano/library/d3viz/index.html).

File: keras/engine/topology.py [hunk 2]
```diff
@@ -464,69 +464,72 @@ def __call__(self, x, mask=None):
             x: can be a tensor or list/tuple of tensors.
             mask: tensor or list/tuple of tensors.
         '''
-        if not self.built:
+        import tensorflow as tf
+        with tf.name_scope(self.name):
```
Comment (fchollet) [inline keras/engine/topology.py]: This should only wrap the call to `build`, btw, not the entire code section.

**Comment by fchollet** [comment]
No update on this, so we won't merge it. We will however introduce proper op naming on the TF side in the near future.

**Comment by chenjiasheng** [comment]
This feature is so cool, no one else think so?
@jakebian, could you please fix the wrap thing and merge it in?

**Comment by hgaiser** [comment]
May I ask what the status is of this PR? Has it been superseded by Keras 2?

**Comment by chenjiasheng** [comment]
@hgaiser
This feature is already merged in and works well for me.
To make tensorboard graph prettier, just name the layers when creating them, and wrap up the relevant code block with `with K.name_scope('name_scope')`.

Check this example, keep eyes on the comments and ignore other things.

```
def ctc_model(model_fn, model_params, n_gpu=1, batch_size=32, use_warpctc=False):
    ### wrap up the origin tower model
    with K.name_scope('Weights'):
        tower = model_fn(**model_params)
    num_features = model_params['num_features']

    ctc_lambda_func = ctc_utils.warpctc_lambda_func if use_warpctc else ctc_utils.ctc_lambda_func
    ### name Lambda layers
    ctc = Lambda(ctc_lambda_func, output_shape=(1,), name="CTCLayer")

    losses = []
    model_inputs = []

    for i in range(n_gpu):
        ### name each tower
        with K.name_scope('Tower'+str(i)):
            if use_warpctc:
                ### name each Input layer
                tower_inputs = [
                    Input(name='TowerInputs' + str(i), shape=(None, num_features)),
                    Input(name='TowerInputsLength' + str(i), shape=(1,), dtype='int32'),
                    Input(name='TowerLabels' + str(i), shape=(None,), dtype='int32'),
                    Input(name='TowerLabelsLengths' + str(i), shape=(1,), dtype='int32')
                ]
            else:
                tower_inputs = [
                    Input(name='TowerInputs' + str(i), shape=(None, num_features)),
                    Input(name='TowerInputsLength' + str(i), shape=(1,), dtype='int32'),
                    Input(name='TowerLabels' + str(i), shape=(None,), dtype='int32', sparse=True)
                ]
            model_inputs += tower_inputs

            import tensorflow as tf
            with tf.device('/gpu:' + str(i)):
                activation = tower(tower_inputs[0])
                loss = ctc([activation] + tower_inputs[1:])
                losses.append(loss)

    if n_gpu > 1:
        ### name the loss layer
        total_loss = average(losses, name='CTCLoss')
    else:
        total_loss = losses[0]

    model = Model(model_inputs, total_loss)
    model.add_loss(total_loss)
    return model


```

**Comment by hgaiser** [comment]
> @hgaiser
> This feature is already merged in and works well for me.
> To make tensorboard graph prettier, just name the layers when creating them, and wrap up the relevant code block with with K.name_scope('name_scope').
> 
> Check this example, keep eyes on the comments and ignore other things.

Thanks, that is more or less what I had too. I am visualing ResNet50, but I get a huge amount of `clip_by_value_{x}` and `Variable_{x}` nodes (where `{x}` is a number). Any idea where these come from? I'm using `name_scope`, similarly to your case.

Perhaps this is out of bounds for this PR though, I should ask this question elsewhere.

**Comment by jakebian** [comment]
Yeah you could always use name scopes explicitly, the point of this PR is to generate these name scopes by default when keras creates layers, so that you don't have to go in and declare 30 namespaces that carefully compartmentalize everything correctly. For example I tested this PR on a resnet50 and it looked clean.

Haven't looked at this for a while, I put it on hold because doing consistent name scoping that matches this in Theano is not trivial. Will have another look soon.

**Comment by sirius0503** [comment]
How can I have a clean visualizaton of a pretrained model, since I wouldn't be able to define `names` and `namescopes`, for the `conv2D` or `Dense` layers in this case. But, I definitely want a cleaner view.

**Pull request closed without merging**